src/order/jordan_hoelder.lean

```diff
@@ -0,0 +1,794 @@
+Jordan Hölder lattice is a lattice equipped with a notion of maximality, `is_maximal`, and a notion
+of isomorphism of pairs `iso`. In the example of subgroups of a group, `is_maximal H K` means that
+`H` is a maximal normal subgroup of `K`, and `iso (H₁, K₁) (H₂, K₂)` means that the quotient
+`H₁ / K₁` is isomorphic to the quotient `H₂ / K₂`. `iso` must be symmetric and transitive and must
+satisfy the second isomorphism theorem `iso (H, H ⊔ K) (H ⊓ K, K)`.
+Examples include `subgroup G` if `G` is a group, and `submodule R M` if `M` is an `R`-module
+-/
+class jordan_hoelder_lattice (X : Type u) [lattice X] :=
+(is_maximal : X → X → Prop)
+(lt_of_is_maximal : ∀ {x y}, is_maximal x y → x < y)
+(sup_eq_of_is_maximal : ∀ {x y z}, is_maximal x z → is_maximal y z →
+  x ≠ y → x ⊔ y = z)
+(is_maximal_inf_left_of_is_maximal_sup : ∀ {x y}, is_maximal x (x ⊔ y) → is_maximal y (x ⊔ y) →
+  is_maximal (x ⊓ y) x)
+(iso : (X × X) → (X × X) → Prop)
+(iso_symm : ∀ {x y}, iso x y → iso y x)
+(iso_trans : ∀ {x y z}, iso x y → iso y z → iso x z)
+(second_iso : ∀ {x y}, is_maximal x (x ⊔ y) → iso (x, x ⊔ y) (x ⊓ y, y))
+
+namespace jordan_hoelder_lattice
+
+variables {X : Type u} [lattice X] [jordan_hoelder_lattice X]
+
+lemma is_maximal_inf_right_of_is_maximal_sup {x y : X}
+  (hxz : is_maximal x (x ⊔ y)) (hyz : is_maximal y (x ⊔ y)) :
+   is_maximal (x ⊓ y) y :=
+begin
+  rw [inf_comm],
+  rw [sup_comm] at hxz hyz,
+  exact is_maximal_inf_left_of_is_maximal_sup hyz hxz
+end
+
+lemma is_maximal_of_eq_inf (x b : X) {a y : X}
+  (ha : x ⊓ y = a) (hxy : x ≠ y)
+  (hxb : is_maximal x b) (hyb : is_maximal y b) :
+  is_maximal a y :=
+begin
+  have hb : x ⊔ y = b,
+    from sup_eq_of_is_maximal hxb hyb hxy,
+  substs a b,
+  exact is_maximal_inf_right_of_is_maximal_sup hxb hyb
+end
+
+lemma second_iso_of_eq {x y a b: X} (hm : is_maximal x a) (ha : x ⊔ y = a) (hb : x ⊓ y = b) :
+  iso (x, a) (b, y) :=
+by substs a b; exact second_iso hm
+
+lemma iso_refl_of_is_maximal {x y : X} (h : is_maximal x y) : iso (x, y) (x, y) :=
+second_iso_of_eq h
+  (sup_eq_right.2 (le_of_lt (lt_of_is_maximal h)))
+  (inf_eq_left.2 (le_of_lt (lt_of_is_maximal h)))
+
+end jordan_hoelder_lattice
+
+open jordan_hoelder_lattice
+
+attribute [symm] iso_symm
+attribute [trans] iso_trans
+
+/--
+A `composition_series X` is a finite nonempty series of elements of a
+`jordan_hoelder_lattice` such that each element is maximal inside the next. The length of a
+`composition_series X` is one less than the number of elements in the series.
+Note that there is no stipulation that a series start from the bottom of the lattice and finish at
+the top. For a composition series `s`, `s.top` is the largest element of the series,
+and `s.bot` is the least element.
+-/
+structure composition_series (X : Type u) [lattice X] [jordan_hoelder_lattice X] : Type u :=
+(length : ℕ)
+(series : fin (length + 1) → X)
+(step' : ∀ i : fin length, is_maximal (series i.cast_succ) (series i.succ))
```

Comment:
Did you consider using lists and https://leanprover-community.github.io/mathlib_docs/data/list/defs.html#list.chain' ?

Reply:
Yes. It might be simpler that way, and I even started refactoring that way, but I don't think it's enough of a difference to justify a refactor, particularly if there is the API to translate it to lists.

Reply:
Ok, fair enough.